rfc/0022-policy-lifecycle.md

```diff
@@ -0,0 +1,397 @@
+
+## Concepts
+
+To implemenet the new policy lifecycle, we will introduce the following concepts:
+
+### Hot reload
+
+The Policy Server will be able to load a new policy without requiring a restart.
+This is will be achieved by refactoring the `EvaluationEnvironment` adding methods to set, update and remove policies.
+
+### The Policy Server acts as a controller
+
+At the time of writing the Policy Server is configured with a ConfigMap that contains all the policies to be loaded.
+When a new policy is created, the ConfigMap is updated, and the Policy Server is restarted to load the new policy.
+
+We want to change this behavior to allow the Policy Server to act as a controller that can load new policies without requiring a restart.
+
+#### PolicyRevision CRD
+
+We will introduce a new CRD called `PolicyRevision` to store the state of a policy in relation to the Policy Server replicas.
+`PolicyRevisions` are namespaced resources, and they will be created in the same namespace as the controller.
+
+The PolicyRevision will include fields to store the policy content and an `enabled` field to specify whether the policy should be served by the Policy Server.
+The `PolicyRevision` will include a `policyGeneration` field to store the generation of the associated `Policy`.
+This field will be set to the value of `metadata.generation` from the `Policy` CRD at the time the `PolicyRevision` is created.
+This approach allows the Policy Server to manage multiple generations of a policy at the same time, facilitating the gradual rollout of updates.
+
+This is similar to the `Deployment` and `ReplicaSet` relationship in Kubernetes, where the `Deployment` defines the desired state at a higher level of abstraction,
+and the `ReplicaSet` is responsible for managing the current state to align with this desired state.
+Additionally, `ReplicaSets` also store the backup of older `Deployment` revisions, enabling rollback to a previous state if needed.
+The `PolicyRevision` will also store the older policy to a maximum of `n` revisions, where `n` is a configurable value.
+This will allow users to rollback to a previous policy version if needed.
+
+In summary:
+
+- A `Policy` specifies the desired state of a policy.
+- One or more `PolicyRevisions` with `enabled` set to `true` represent the active policies currently served by the Policy Server.
+- A `PolicyRevision` with `enabled` set to `true` and `policyGeneration` matching the current generation of the `Policy` represents the desired state of the policy within the Policy Server's evaluation environment.
+- `PolicyRevisions` with `enabled` set to `false` represent previous states of the policies, retained for potential rollback scenarios.
+
+When a new Policy is created, the controller will create a new `PolicyRevision` resource with the policy content and set the `enabled` field to `true`.
```

Comment:
the controller here means the policy-server, correct? As in, only policy-servers are controllers for Policies and PolicyRevisions?

Reply:
the controller means the main golang controller. even if the policy server acts as a controller for policy revision, the first creation is made by the kw controller.